### ISSUE TITLE
RFC: streamlining the boilerplate

### ISSUE THREAD

**Issue by dherman** [neon-bindings/neon-cli#21]
With @eddyb's work in rustbridge/neon#29, I can see a glimpse of a more streamlined project structure for Neon projects. I'll probably get working on a branch ASAP but I thought I'd put down my thoughts here to get feedback, and also as a heads up to folks who've been helping get our infrastructure started.
# Overview

There are three major components to the streamlined structure:
- eliminating the dependence on node-gyp, which means no more `binding.cc` and no more `binding.gyp`
- move the native directory structure into a `native/` subdirectory
- eliminate rust-bridge and return to having a `neon build` command that's used by `package.json`

The basic project structure would look like this:

```
foobar/
├── package.json
├── lib/
│   └── index.js
└── native/
    ├── Cargo.toml
    └── src/
        └── lib.rs
```

After building, a project would look like:

```
foobar/
├── package.json
├── lib/
│   └── index.js
└── native/
    ├── Cargo.toml
    ├── Cargo.lock
    ├── src/
    │   └── lib.rs
    ├── target/
    │   └── release/
    │       └── libfoobar.so
    └── index.node
```

Within the project, loading the native module (for example from a module in the `lib/` directory) would look like:

``` javascript
var nmod = require('../native');
```

The `package.json` would require a postinstall hook:

``` javascript
"scripts": {
  "postinstall": "neon build"
}
```
# Details
## Eliminating dependency on node-gyp

The vast majority of the required C++ code is used in node-sys to create the actual Neon library itself. But for a Neon project, the only remaining C++ code was the tiny bit of top-level boilerplate code in `binding.cc`, which @eddyb has shown in rustbridge/neon#29 we can do in pure Rust. This means we don't need node-gyp at all for building an individual project, and we can eliminate both `binding.cc` and `binding.gyp` entirely. (The node-sys project would continue to be built using node-gyp.)
## Create a `native/` subdirectory

This has several benefits:
- **clearly puts NPM in the driver's seat**: A Neon project is, from the outside world's perspective, an NPM package. The Rust implementation is internal to the project. Putting the native code in a subdirectory makes it clear that the outer structure of the project is the standard structure of an NPM project.
- **no collision with Node `src` directory**: The convention for compile-to-JS (including ES6 via Babel) is to put it in `src/`, which creates a collision with having the Cargo project and the NPM project share the same root directory. Moving the Cargo project into `native/` eliminates the overlap.
- **still allows zero JS code**: A Neon project that's completely written in Rust can simply set its package.json's main entry point to `native/index.node` and not have a `lib/` directory at all.
## Add `neon build` back to `package.json`

Having the build step go through our own entry point puts us in control of the build process, and allows us to script it in JS instead of GYP (which has proven to be extremely challenging). The basic logic of the build step is:
- run cargo, potentially with custom linker args (e.g. on OS X, which needs extra arguments to tell it to allow creating a dylib with undefined symbols)
- symlink (on Windows, maybe copy? :-/) the output to `native/index.node`
  Having our own build step will also allow us to do more complex things in the future, like have logic for fetching prebuilt binaries instead of building from scratch.
# Drawbacks

I was temporarily attracted to the idea of eliminating a dependency on `neon-cli` in a Neon project's `package.json`. However, a) there's no net increase in boilerplate within the `package.json`, since we eliminate the dependency on neon-bridge; b) it's not unreasonable to have a dependency on a build tool in a Node package's build step; c) having a permanent dependency on node-gyp for kicking off all build logic is just painful to contemplate.

/cc @mmun, @watilde

**Comment by dherman** [comment]
note before I forget: https://www.npmjs.com/package/symlink-or-copy

**Comment by dherman** [comment]
> note before I forget: https://www.npmjs.com/package/symlink-or-copy

Turns out Node barfs if you load a .node symlink whose target is not also a .node filename, so symlinking is out; we'll just unconditionally copy no matter what OS.

**Comment by dherman** [comment]
Landed in v0.1.6.
